Automattic/wp-calypso
# Promotions: Add all validations and empty checks

This adds all remaining validations and also checks for empty values
before submitting to save or update.

It also changes the behavior of the Save/Update button. The button is
always enabled, but displays an error if any validation errors are
present. When pressed, it also enables "empty" validation errors to be
displayed. These are hidden from the user on the initial loading of the
page.

Design Notes:
* Please look over copy as there's a lot of improvised text here. Feel free to add a commit to change copy, no need to check with me.
* There should be a lot of versatility to change things here if it's needed, so feel free to discuss  if you feel things should operate differently.

Code Notes:
* This PR is a bit bigger than I had planned (sorry!), but includes an entirely new file for validation functions, which should be straightforward to review.
* It also adjusts the structure of the model objects to accommodate both a component to render and a validation function for each. This helps keep each field's context near each other.

To Test:
From the "Add Promotion" page:
* Try clicking "Save", observe error and "empty" validation errors popping up.
* Try adding conditions with empty fields and saving.
* With a product sale, try setting the start date after the end date and saving.
* Try setting a min amount larger than a max amount.
* Try setting a per-user limit larger than a total limit.
* Try setting any numeric field to 0 or negative.

From the "Update Promotion" page:
* Try clicking "Save" without changing anything. It should work and do nothing.
* Try the above tests, same as "Add Promotion" page.

![new_promotion_ _coderkevin_ _wordpress_com](https://user-images.githubusercontent.com/945228/33060409-bafb904e-ce5d-11e7-8da7-fbbcbe594f45.png)

![edit_promotion_ _coderkevin_ _wordpress_com](https://user-images.githubusercontent.com/945228/33060376-9f029798-ce5d-11e7-814a-353a01d3c767.png)
